### PR TITLE
Multisample seurat

### PR DESCRIPTION
**Pull request by JuanluOnieva-biomage**
Modify the data ingest script to filter, process and load multiple-sample data and metadata into the platform

Link for jira ticket: https://biomage.atlassian.net/browse/BIOMAGE-392?atlOrigin=eyJpIjoiNDBlNzhhMGEyY2FhNDE5NmI4ZTVjOWJlNDBkOGI0MTciLCJwIjoiaiJ9

- Information about metadata should be describe in meta.json. 
- Possible structure of config meta.json

{
    "name": "name of the experiment",
    "organism": "hsapiens/ggallus/mmusculus",
        "samples": {
    	"multisample": "FALSE/TRUE",
    	"samples_info":{
    		"type": [ _vector with the name of the samples_ ]
                 _...we can add here more metadata information..._
    	}
    },
    "input": {
        "type": "10x"
    }
}

Also, this PR contain the  embedding calculation with the default setting for the jira ticket https://biomage.atlassian.net/browse/BIOMAGE-488?atlOrigin=eyJpIjoiN2MxNThiZTYyN2MyNDVjOWEzNDcyM2VkYzVkYmFiOWUiLCJwIjoiaiJ9 and persists the MT content and the doublet scores.

### CODE REVIEW

File: src/1-preproc.r [hunk 1]
```diff
@@ -200,10 +215,10 @@ message("Creating raw dataframe...")
 data <- create_dataframe(config)
```
Comment (seb-mueller) [inline src/1-preproc.r:216]: `str(data)` gives my this for the multisample dataset, is this about right?
```
List of 1
 $ raw:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. ..@ i       : int [1:15960019] 36 61 67 115 116 128 151 172 186 197 ...
  .. ..@ p       : int [1:11268] 0 1548 3580 5275 6755 8559 10021 10424 11938 13487 ...
  .. ..@ Dim     : int [1:2] 31053 11267
  .. ..@ Dimnames:List of 2
  .. .. ..$ : chr [1:31053] "ENSMUSG00000051951" "ENSMUSG00000089699" "ENSMUSG00000102343" "ENSMUSG00000025900" ...
  .. .. ..$ : chr [1:11267] "KO_AAACCCAAGAGCCTGA-1" "KO_AAACCCAAGATGACCG-1" "KO_AAACCCACAACTGTGT-1" "KO_AAACCCATCGTAACCA-1" ...
  .. ..@ x       : num [1:15960019] 1 2 2 1 1 1 2 1 1 1 ...
  .. ..@ factors : list()
```

File: src/1-preproc.r [hunk 2]
```diff
@@ -200,10 +215,10 @@ message("Creating raw dataframe...")
...
 
 message("Filtering cells by size")
-data$filtered <- data$raw[, Matrix::colSums(data$raw)>=1e3]
+data$filtered <- data$raw[, Matrix::colSums(data$raw>0)>=200]
```
Comment (seb-mueller) [inline src/1-preproc.r:218]: Those are hardcoded thresholds, we should probably discuss if there is a better way to deal with this. It's fine for the time being. 
However, why did you change it? Do you have any sources which you base those threshols on?

**Review by seb-mueller — COMMENTED** [review]
This is the first time I've been reading the code, so I have a few general but also specific remarks, which we should probably discuss in a meeting. 

- There is a lot of hardcoded thresholds and parameters, this should probably be written in a config file instead
- it was not obvious to me where code snippets are from, i.e. borrowed from other sources or self-written. For the latter we have to be careful since it is not tested as much as public code, so we need to flag our sources!
- We should use a lot more comments in general!
- I was a bit suprised that there is not code that can deal with reading in multisample elsewhere. How are the Seurat guys going about it?

File: src/3-preproc.r [hunk 3]
```diff
@@ -3,117 +3,184 @@ library(Matrix)
...
-message("Storing annotations...")
-#Misc(data, slot="annotations") <- annotations
-data@misc$gene_annotations <- annotations
-
 message("Adding MT information...")
-data <- PercentageFeatureSet(data, pattern = "^MT-", col.name = "percent.mt")
+data_gene_symbol <- data
+rownames(data_gene_symbol@assays$RNA@data) <- annotations$name[match(rownames(data), annotations$input)]
+rownames(data_gene_symbol@assays$RNA@counts) <- annotations$name[match(rownames(data), annotations$input)]
+data_gene_symbol <- PercentageFeatureSet(data_gene_symbol, pattern = "^MT-", col.name = "percent.mt")
```
[the remark this answers is not in the document]
Comment (JuanluOnieva-biomage) [reply src/3-preproc.r]: You are right. With gProfiler we can only know the MT genes from mouse and humans, I have checked. So for the rest of the organism we need to think about it.

File: src/1-preproc.r [hunk 4]
```diff
@@ -146,30 +150,41 @@ Read10X_data <- function (data.dir = NULL, gene.column = 2, unique.features = TR
...
-    message(
-        paste(
-            "Found", nrow(data$raw), "genes and", ncol(data$raw), "cells."
-        )
+  data_type <- config$input["type"]
+  data <- list()
+  
+  #Check config format
+  if(!"samples"%in%names(config))
+    stop("The format of the config is wrong. There should be a category for the sample.")
```
[the remark this answers is not in the document]
Comment (JuanluOnieva-biomage) [reply src/1-preproc.r]: Done!

File: src/1-preproc.r [hunk 5]
```diff
@@ -146,30 +150,41 @@ Read10X_data <- function (data.dir = NULL, gene.column = 2, unique.features = TR
...
-
-        data$raw <- as.matrix(read.table(paste("/input", path, sep = "")))
-    }
-
-    message(
-        paste(
-            "Found", nrow(data$raw), "genes and", ncol(data$raw), "cells."
-        )
+  data_type <- config$input["type"]
+  data <- list()
```
[the remark this answers is not in the document]
Comment (JuanluOnieva-biomage) [reply src/1-preproc.r]: Yes... in the previous script it was used so I keep it. Since we do not load anything I do not worry about it, nevertheless I have refactor it!

File: src/1-preproc.r [hunk 6]
```diff
@@ -146,30 +150,41 @@ Read10X_data <- function (data.dir = NULL, gene.column = 2, unique.features = TR
...
+  
+  #Check config format
+  if(!"samples"%in%names(config))
```
[the remark this answers is not in the document]
Comment (JuanluOnieva-biomage) [reply src/1-preproc.r:157]: Yes, but I think that this is compatible with testhat framework

File: src/1-preproc.r [hunk 7]
```diff
@@ -14,47 +14,59 @@ library(MASS)
...
 #' @param gene.column gene column number
 #' @param unique.features not duplicated features
 #' @param strip.suffix boolean to remove -1 at the end of barcodes (default FALSE)
+#' @param sample in case of multiplesample, it refers to the name of the samples
 #' 
-#' @return dgCMatrix with genes as rows and cells as column
+#' @return dgCMatrix with genes as rows and cells as column. In the case of multisample,
+#'  the cell names have a prefix _ corresponding to the type metadata from the json.
 
 Read10X_data <- function (data.dir = NULL, gene.column = 2, unique.features = TRUE,
```
[the remark this answers is not in the document]
Comment (JuanluOnieva-biomage) [reply src/1-preproc.r:27]: The think is that this function "Read10X_data" belongs to Seurat. Since I did some modifications I prefer to leave it with the same name. I just add the cite (I forgot to do it). I have only comment the part of the code that I write by my-self in this function. But If you think it's worth it, I can comment the Seurat part of this function

File: src/1-preproc.r [hunk 8]
```diff
@@ -65,20 +77,12 @@ Read10X_data <- function (data.dir = NULL, gene.column = 2, unique.features = TR
       cell.names <- as.vector(x = as.character(x = sapply(X = cell.names, 
                                                           FUN = ExtractField, field = 1, delim = "-")))
     }
```
[the remark this answers is not in the document]
Comment (JuanluOnieva-biomage) [reply src/1-preproc.r:79]: Because it belongs to seurat env

File: src/1-preproc.r (hunk 1 above)
Comment (JuanluOnieva-biomage) [reply src/1-preproc.r:218]: Yes it is!

Do you miss anything?

File: src/1-preproc.r (hunk 2 above)
Comment (JuanluOnieva-biomage) [reply src/1-preproc.r]: They are the deafult values when converting to seurat object. But I put them here becuase if I run scrublet directly without these filters, it brokes because the amount of empty information. Because of that I have included here. 

Regardales a possible json with the default values, the problem is that in the case of the tsne, the default values depends on the number of cells. 

I know that when the data-ingest is done, this config information is stored in dynamodb. Nevertheless, it could be a good idea to have them from the beggining in a json, we can discuss the benefit of it!

**Comment by JuanluOnieva-biomage** [comment]
Hi @seb-mueller I have did some comments in each review. Regarding your questions and points:

1. Yes it could be a good solution. But in the case of tsne, the parameter depends on the number of cells. We should discuss the benefits of saving then in a file or not. 
2. You are right! I have cited the other source code. 
3. Normally, the self-written code is commented. But feel free to ping the part of the code and I will add more comments! sorry for that 
4. There is, the original Read10X_data supports multisample but without flexibility in terms of input folder, that is, file name and folder structure.

**Comment by seb-mueller** [comment]
> Hi @seb-mueller I have did some comments in each review. Regarding your questions and points:
> 
>     1. Yes it could be a good solution. But in the case of tsne, the parameter depends on the number of cells. We should discuss the benefits of saving then in a file or not.
Ok, I think for this should do for the time being then. But yes, let's discuss this in a meeting

>     2. You are right! I have cited the other source code.
Thanks, publications cites are good, but was more referring to the original source of code snippets.
E.g. as in 4) the `Read10X_data` seems based on `Seurat::Read10X` function (https://satijalab.org/seurat/reference/read10x)? Is that right. As you said in 4) there is a reason for it, so this should be stated in the comments.


>     3. Normally, the self-written code is commented. But feel free to ping the part of the code and I will add more comments! sorry for that
Thanks! Maybe, as a suggestion, I could go through the code myself and add comments, so this way I also get to understand the bits. This doesn't have to be in this PR since it seems urgent.
 
>     4. There is, the original Read10X_data supports multisample but without flexibility in terms of input folder, that is, file name and folder structure.
I see. Since I first looked at the code and now understand the reason for this.
In general, I think we should avoid copying and modify code without a very a good reason, since we have to maintain our own codebase and can't benefit from improvents of the orignal source!
In this case, the reason is not very convincing to me, i.e. I suspect this is because `Read10X` expects a folder for each sample which is not the case for our test data. But do we know this is the case for other data as well?
Do we have any other multi-sample data sets? How is CellRangers out put for multi-samples anyway?
We should probably talk about this in a meeting if the answers are unclear.

File: src/1-preproc.r (hunk 7 above)
[the remark this answers is not in the document]
Comment (seb-mueller) [reply src/1-preproc.r:27]: I understand now :) I've addressed this point below for a more general discussion about this point.

File: src/1-preproc.r (hunk 6 above)
[the remark this answers is not in the document]
Comment (seb-mueller) [reply src/1-preproc.r:157]: Ok, we can discuss this this more in general with Oli, it's good for now anyway!

File: src/1-preproc.r (hunk 1 above)
Comment (seb-mueller) [reply src/1-preproc.r:218]: No, just checking :)

File: src/1-preproc.r (hunk 2 above)
Comment (seb-mueller) [reply src/1-preproc.r]: I see, yes, we have to discuss this with the engineers in more detail. 
Could you add the justification you wrote above in the comments of the code?

File: src/3-preproc.r [hunk 9]
```diff
@@ -3,117 +3,184 @@ library(Matrix)
...
-
-#In order to determine the filter parameters for mito genes we would need to plot them on either a violin plot or a scatter plot. Please refer to the plots above.
+message("Loading configuration...")
+config <- RJSONIO::fromJSON("/input/meta.json")
+metadata <- check_config(data, config)
 
-#Remove by subsetting... or look under ScaleData below for regressing out mito genes.
-#data <- subset(data, subset = percent.mt < 5)
+message("Creating Seurat Object...")
+data <- Seurat::CreateSeuratObject(data$filtered, assay='RNA', min.cells=3, min.features=200, meta.data=metadata)
```
Comment (seb-mueller) [inline src/3-preproc.r]: `data` naming again. We should probably name it `seuratobj` or so instead

File: src/2-run_scrublet.py [hunk 10]
```diff
@@ -1,12 +1,35 @@
...
 
-scrub = scr.Scrublet(df)
-doublet_scores, _ = scrub.scrub_doublets()
+with open('/input/meta.json') as f:
+    config = json.load(f)
+
+# Checking multisample
+if config['samples']['multisample'] == 'TRUE':
+  samples = config['samples']['samples_info']['type']
+  # Running first sample
```
Comment (ogibson-biomage) [inline src/2-run_scrublet.py:16]: Why run the first one separate from the rest?

File: src/3-preproc.r [hunk 11]
```diff
@@ -3,144 +3,211 @@ library(Matrix)
...
+#' @param scdata matrix with barcodes as columns to assign metadata information
+#' @param config config list from meta.json
+#' 
+#' @export save barcodes to keep
+
+check_config <- function(scdata, config){
+    metadata <- NULL
+    
+    if("type" %in% names(config$samples$samples_info)){
+        metadata <- data.frame(row.names = colnames(scdata$filtered))
```
Comment (ogibson-biomage) [inline src/3-preproc.r:38]: Some comments here would be very helpful to make the code a bit more readable!

File: src/3-preproc.r [hunk 12]
```diff
@@ -3,144 +3,211 @@ library(Matrix)
...
+}else{
+    scdata <- Seurat::NormalizeData(scdata, normalization.method = normalization_method, scale.factor = normalization_scale_factor, verbose = F)
+    scdata <-Seurat::FindVariableFeatures(scdata, selection.method = "vst", nfeatures = nfeatures, verbose = F)
+    scdata<-Seurat::ScaleData(scdata, verbose = F)
+    
+    vars <- HVFInfo(object = scdata, assay = "RNA", selection.method = 'vst')
+}
 
-vars <- HVFInfo(object = data, assay = "RNA", selection.method = 'vst')
 vars <- vars[, 'variance.standardized', drop=FALSE]
```
Comment (ogibson-biomage) [inline src/3-preproc.r:127]: Did you mean storing this vars inside the rds? Just as an idea, we could calculate variable features with nfeatures = maxNfeatures, then calculate HVFInfo and recalculate the variable features.

Comment (ogibson-biomage) [reply src/3-preproc.r:127]: There must be an easier way of doing this, though

File: src/1-preproc.r (hunk 6 above)
[the remark this answers is not in the document]
Comment (ogibson-biomage) [reply src/1-preproc.r:157]: Hmm, I was thinking of using it to check Seurat objects and see if they are platform compatible as well, but I don't see what we could do with it here

File: src/1-preproc.r (hunk 2 above)
Comment (ogibson-biomage) [reply src/1-preproc.r]: A few things I'd like to comment on this. 

I do agree that we should use library functions whenever possible instead of tweaking them because it saves a lot of time. If we can solve the input folder structure problem in a different way that'd be really great. 

I don't see the benefit of a json file for default config parameters. We can use dynamoDb (that basically stores json files) for everything config-related. And in the case of parameters that are dependant on the dataset i.e tsne parameters, then we can calculate those and add them to dynamo on the ingest script. I know we should discuss the default parameters, but I think using those set by Seurat is a wise choice that can be easily defended.

That being said, I think approving this MR is a bit urgent so I think we could leave these issues for tomorrow's bioinfo meeting because they don't seem changes that could potentially cause problems.

Comment (JuanluOnieva-biomage) [reply src/1-preproc.r]: I have added the original source of the Seurat parameter

File: src/2-run_scrublet.py (hunk 10 above)
Comment (JuanluOnieva-biomage) [reply src/2-run_scrublet.py]: There is no reason... so I simplify the code now :)

File: src/3-preproc.r (hunk 11 above)
Comment (JuanluOnieva-biomage) [reply src/3-preproc.r:41]: I hope now is more clear

File: src/3-preproc.r (hunk 12 above)
Comment (JuanluOnieva-biomage) [reply src/3-preproc.r:127]: I save the dispersion inside misc as gene_dispersion

**Review by seb-mueller — APPROVED** [review]
Just had a call with Juanlu and approved the changes
